### PR TITLE
Support thread command by gdb python script

### PR DESCRIPTION
**Pull request by anjiahao1**
## Summary
Support thread command on nuttx

support arch:
  sim x86
  sim x86_64
  arm32 all series

  If we need to support more architectures in the future,
  just add the 'g_tcbinfo' of the corresponding architecture

support thread command:
  1. thread <id>
    switch thread
  2. info thread
    list all thread info
  3. thread apply [all | id list] command
    like this:
      thread apply all bt
      thread apply all bt full
      thread apply 1 2 4 bt
      thread apply all info r
  4. nxsetregs
      Set registers to the specified values.
      Usage: nxsetregs [regs]

      Etc:
           nxsetregs
           nxsetregs g_current_regs[0]
           nxsetregs tcb->xcp.regs
           Nxsetregs g_pidhash[0].tcb->xcp.regs

      Default regs is g_current_regs[0],if regs is NULL,
      it will not set registers.

      Because NuttX enters exception_common during a crash,
      this assembly function manipulates the stack pointer (sp),
      causing GDB to be unable to trace back to the first context of the crash.
      Instead, it shows the context of the interrupt stack. By using nxsetregs,
      it is possible to forcefully set the first context to
      the one before the interrupt occurred.

Notice:
  Switching threads is achieved by setting registers,
  so registers need to be restored before continue,
  please use 'c' to continue instead of 'continue'

## Impact
gdb debug
## Testing
sim32
sim64
stm32f429i-disco with jlink

## Usage
 CONFIG_DEBUG_TCBINFO=y

add `'-ex "souce tools/gdb/__init__.py"'` in gdb command or `'souce tools/gdb/__init__.py'` in gdb shell

### CODE REVIEW

File: tools/gdb/thread.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+        super(Nxcontinue, self).__init__("nxcontinue", gdb.COMMAND_USER)
+
+    def invoke(self, args, from_tty):
+        restore_regs()
+        gdb.execute("continue")
+
+
+# We can't use a user command to rename continue it will recursion
+gdb.execute("define c\n nxcontinue \n end\n")
+gdb.write("\nif use thread command,plase don't use 'continue', use 'c' inside !!!\n")
```
Comment (xiaoxiang781216) [inline tools/gdb/thread.py:265]: add space after ,

File: tools/gdb/thread.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+                i = 0
+                cmd = ""
+                for i in range(1, arglen):
+                    if arg[i].isnumeric():
+                        threadlist.append(int(arg[i]))
+                    else:
+                        cmd += arg[i] + " "
+
+                if len(threadlist) == 0 or cmd == "":
+                    gdb.write("Please specify a thread ID list and command\n" % cmd)
```
Comment (xiaoxiang781216) [inline tools/gdb/thread.py:227]: remove "% cmd"

File: tools/gdb/thread.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+            if g_tcbinfo["reg_off"]["p"][i] != UINT16_MAX:
+                value = gdb.Value(regs + g_tcbinfo["reg_off"]["p"][i]).cast(
+                    gdb.lookup_type("uintptr_t").pointer()
+                )[0]
+                gdb.execute("set $%s = 0x%x" % (reg.name, value))
+
+            i += 1
+
+        if current_regs[0] == 0:
```
[the remark this answers is not in the document]
Comment (anjiahao1) [reply tools/gdb/thread.py:110]: move to 85

File: tools/gdb/thread.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        print("%-4s %-21s %-80s %-30s" % ("Id", "Thread", "Info", "Frame"))
+        for i in range(0, g_npidhash):
+            if g_pidhash[i] == 0:
+                continue
+
+            if g_pidhash[i]["task_state"] == 3:
+                gdb.write("*")
+                pc = int(gdb.parse_and_eval("$pc"))
+            else:
+                gdb.write(" ")
```
Comment (xiaoxiang781216) [inline tools/gdb/thread.py:147]: id = "%s" % i

File: tools/gdb/thread.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        g_npidhash = gdb.parse_and_eval("g_npidhash")
+        g_pidhash = gdb.parse_and_eval("g_pidhash")
+        g_statenames = gdb.parse_and_eval("g_statenames")
+        print("%-4s %-21s %-80s %-30s" % ("Id", "Thread", "Info", "Frame"))
+        for i in range(0, g_npidhash):
+            if g_pidhash[i] == 0:
+                continue
+
+            if g_pidhash[i]["task_state"] == 3:
+                gdb.write("*")
```
Comment (xiaoxiang781216) [inline tools/gdb/thread.py:144]: id = "*%s" % i

File: tools/gdb/thread.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+                pc = int(gdb.parse_and_eval("$pc"))
+            else:
+                gdb.write(" ")
+                pc = get_pc_value(g_pidhash[i])
+
+            id = "%s" % i
```
Comment (xiaoxiang781216) [inline tools/gdb/thread.py:150]: remove

File: tools/gdb/thread.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+# under the License.
+#
+############################################################################
+
+import gdb
+import utils
+
+UINT16_MAX = 0xFFFF
+
+g_tcbinfo = gdb.parse_and_eval("g_tcbinfo")
```
Comment (xiaoxiang781216) [inline tools/gdb/thread.py:26]: let's remove global variables to avoid the cache value stale after user reload image.

Comment (anjiahao1) [reply tools/gdb/thread.py]: ok

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
let ignore xtensa specific ci temp broken.

**Comment by masayuki2009** [comment]
@anjiahao1 

I noticed that ostest with qemu-armv8a:netnsh_smp causes a deadlock.

```
user_main: signal handler test
sighand_test: Initializing semaphore to 0
sighand_test: Unmasking SIGCHLD
sighand_test: Registering SIGCHLD handler
sighand_test: Starting waiter task
sighand_test: Started waiter_main pid=153
waiter_main: Waiter started
waiter_main: Unmasking signal 32
waiter_main: Registering signal handler
waiter_main: oact.sigaction=0 oact.sa_flags=0 oact.sa_mask=0000000000000000
waiter_main: Waiting on semaphore
sighand_test: Signaling pid=153 with signo=32 sigvalue=42
```

If I rever the following commit, the deadlock disappeared.

```
commit 8abf803a1d7147efebc987a8ec5ac961a9a50515
Author: anjiahao <anjiahao@xiaomi.com>
Date:   Fri Jul 21 19:59:59 2023 +0800

    arm64:Need to store native sp pointer to tcb

    Signed-off-by: anjiahao <anjiahao@xiaomi.com>
```

**Comment by masayuki2009** [comment]
@anjiahao1 

>I noticed that ostest with qemu-armv8a:netnsh_smp causes a deadlock.

The ostest crashed with qemu-armv8a:nsh_smp.

```
user_main: signal handler test
sighand_test: Initializing semaphore to 0
sighand_test: Starting waiter task
sighand_test: Started waiter_main pid=155
waiter_main: Waiter started
waiter_main: Unmasking signal 32
waiter_main: Registering signal handler
waiter_main: oact.sigaction=0 oact.sa_flags=0 oact.sa_mask=0000000000000000
waiter_main: Waiting on semaphore
sighand_test: Signaling pid=155 with signo=32 sigvalue=42
[CPU0] nxsig_queue: pid=0x0000009b signo=32 value=42
[CPU0] nxsig_tcbdispatch: TCB=0x40378700 pid=155 signo=32 code=1 value=42 masked=NO
[CPU0] up_schedule_sigaction: tcb=0x40378700 sigdeliver=0x4028b6e0
[CPU0] up_schedule_sigaction: rtcb=0x4037d550 CURRENT_REGS=0
[CPU1] arm64_fatal_error: reason = 1
[CPU1] arm64_fatal_error: CurrentEL: MODE_EL1
[CPU1] arm64_fatal_error: ESR_ELn: 0x1fe00000
[CPU1] arm64_fatal_error: FAR_ELn: 0x0
[CPU1] arm64_fatal_error: ELR_ELn: 0x40290994
[CPU1] print_ec_cause: Trapped access to SVE, Advanced SIMD, or floating-point functionality
[CPU1] _assert: Current Version: NuttX  10.4.0 0092b3e30f Jul 24 2023 19:30:27 arm64
[CPU1] _assert: Assertion failed panic: at file: common/arm64_fatal.c:375 task(CPU1): waiter 0x402a2d24
[CPU1] up_dump_register: stack = 0x40373f30
[CPU1] up_dump_register: x0:   0x402c5ea8          x1:   0xdeaddead
[CPU1] up_dump_register: x2:   0x402bc0d0          x3:   0x40286b98
[CPU1] up_dump_register: x4:   0x402c5ee8          x5:   0x40287c94
[CPU1] up_dump_register: x6:   0x40378700          x7:   0x1
[CPU1] up_dump_register: x8:   0x402bc0d0          x9:   0x40286b98
[CPU1] up_dump_register: x10:  0x402c5ee8          x11:  0x40287c94
[CPU1] up_dump_register: x12:  0x40378700          x13:  0x1
[CPU1] up_dump_register: x14:  0x402c5ee4          x15:  0x40287cfc
[CPU1] up_dump_register: x16:  0x1                 x17:  0x0
[CPU1] up_dump_register: x18:  0x402c5000          x19:  0x40378700
[CPU1] up_dump_register: x20:  0x402c5000          x21:  0x40286f94
[CPU1] up_dump_register: x22:  0x402bc150          x23:  0x0
[CPU1] up_dump_register: x24:  0x40378bac          x25:  0x402bc150
[CPU1] up_dump_register: x26:  0x0                 x27:  0x4028c2ec
[CPU1] up_dump_register: x28:  0x402bc150          x29:  0x402bc0d0
[CPU1] up_dump_register: x30:  0x40378bac
[CPU1] up_dump_register:
[CPU1] up_dump_register: STATUS Registers:
[CPU1] up_dump_register: SPSR:      0x4008e610
[CPU1] up_dump_register: ELR:       0x402bc168
[CPU1] up_dump_register: SP_EL0:    0x40374060
[CPU1] up_dump_register: SP_ELX:    0x40374060
[CPU1] up_dump_register: TPIDR_EL0: 0x40378700
[CPU1] up_dump_register: TPIDR_EL1: 0x40378700
[CPU1] up_dump_register: EXE_DEPTH: 0x1
[CPU1] dump_tasks:    PID GROUP   CPU PRI POLICY   TYPE    NPX STATE   EVENT      SIGMASK          STACKBASE  STACKSIZE      USED   FILLED    COMMAND
[CPU1] dump_tasks:   ----   ---  ---- --- -------- ------- --- ------- ---------- -------- 0x402cb000      8192      8192   100.0%!   irq
[CPU1] dump_task:       0     0     0   0 FIFO     Kthread N-- Assigned           0000000000000000 0x402d1010     16368      1360     8.3%    CPU0 IDLE
[CPU1] dump_task:       1     1     1   0 FIFO     Kthread N-- Assigned           0000000000000000 0x402d5010     16368      1088     6.6%    CPU1 IDLE
[CPU1] dump_task:       2     2     2   0 FIFO     Kthread N-- Running            0000000000000000 0x402d9010     16368      1088     6.6%    CPU2 IDLE
[CPU1] dump_task:       3     3     3   0 FIFO     Kthread N-- Running            0000000000000000 0x402dd010     16368      1088     6.6%    CPU3 IDLE
[CPU1] dump_task:       4     4     0 192 RR       Kthread --- Waiting Semaphore  0000000000000000 0x402e39f0     16320      1264     7.7%    hpwork 0x402c64f8
[CPU1] dump_task:       5     5     0 100 RR       Task    --- Waiting Semaphore  0000000000000000 0x402e8000     16336      2944    18.0%    nsh_main
[CPU1] dump_task:      22    22     0 100 RR       Task    --- Running            0000000000000000 0x4036c6e0     16288      2320    14.2%    ostest Arg1 Arg2 Arg3 Arg4
[CPU1] dump_task:     155   155     1 100 RR       Task    --- Running            0000000000000000 0x403706c0     16336      3552    21.7%    waiter
[CPU1] dump_task:      15    15     0 100 RR       Task    --- Waiting Semaphore  0000000000000000 0x40379530     16336      1984    12.1%    ostest
```

Could you please confirm if the latest upstream works with qemu?

**Comment by anjiahao1** [comment]
> @anjiahao1
> 
> > I noticed that ostest with qemu-armv8a:netnsh_smp causes a deadlock.
> 
> The ostest crashed with qemu-armv8a:nsh_smp.
> 
> ```
> user_main: signal handler test
> sighand_test: Initializing semaphore to 0
> sighand_test: Starting waiter task
> sighand_test: Started waiter_main pid=155
> waiter_main: Waiter started
> waiter_main: Unmasking signal 32
> waiter_main: Registering signal handler
> waiter_main: oact.sigaction=0 oact.sa_flags=0 oact.sa_mask=0000000000000000
> waiter_main: Waiting on semaphore
> sighand_test: Signaling pid=155 with signo=32 sigvalue=42
> [CPU0] nxsig_queue: pid=0x0000009b signo=32 value=42
> [CPU0] nxsig_tcbdispatch: TCB=0x40378700 pid=155 signo=32 code=1 value=42 masked=NO
> [CPU0] up_schedule_sigaction: tcb=0x40378700 sigdeliver=0x4028b6e0
> [CPU0] up_schedule_sigaction: rtcb=0x4037d550 CURRENT_REGS=0
> [CPU1] arm64_fatal_error: reason = 1
> [CPU1] arm64_fatal_error: CurrentEL: MODE_EL1
> [CPU1] arm64_fatal_error: ESR_ELn: 0x1fe00000
> [CPU1] arm64_fatal_error: FAR_ELn: 0x0
> [CPU1] arm64_fatal_error: ELR_ELn: 0x40290994
> [CPU1] print_ec_cause: Trapped access to SVE, Advanced SIMD, or floating-point functionality
> [CPU1] _assert: Current Version: NuttX  10.4.0 0092b3e30f Jul 24 2023 19:30:27 arm64
> [CPU1] _assert: Assertion failed panic: at file: common/arm64_fatal.c:375 task(CPU1): waiter 0x402a2d24
> [CPU1] up_dump_register: stack = 0x40373f30
> [CPU1] up_dump_register: x0:   0x402c5ea8          x1:   0xdeaddead
> [CPU1] up_dump_register: x2:   0x402bc0d0          x3:   0x40286b98
> [CPU1] up_dump_register: x4:   0x402c5ee8          x5:   0x40287c94
> [CPU1] up_dump_register: x6:   0x40378700          x7:   0x1
> [CPU1] up_dump_register: x8:   0x402bc0d0          x9:   0x40286b98
> [CPU1] up_dump_register: x10:  0x402c5ee8          x11:  0x40287c94
> [CPU1] up_dump_register: x12:  0x40378700          x13:  0x1
> [CPU1] up_dump_register: x14:  0x402c5ee4          x15:  0x40287cfc
> [CPU1] up_dump_register: x16:  0x1                 x17:  0x0
> [CPU1] up_dump_register: x18:  0x402c5000          x19:  0x40378700
> [CPU1] up_dump_register: x20:  0x402c5000          x21:  0x40286f94
> [CPU1] up_dump_register: x22:  0x402bc150          x23:  0x0
> [CPU1] up_dump_register: x24:  0x40378bac          x25:  0x402bc150
> [CPU1] up_dump_register: x26:  0x0                 x27:  0x4028c2ec
> [CPU1] up_dump_register: x28:  0x402bc150          x29:  0x402bc0d0
> [CPU1] up_dump_register: x30:  0x40378bac
> [CPU1] up_dump_register:
> [CPU1] up_dump_register: STATUS Registers:
> [CPU1] up_dump_register: SPSR:      0x4008e610
> [CPU1] up_dump_register: ELR:       0x402bc168
> [CPU1] up_dump_register: SP_EL0:    0x40374060
> [CPU1] up_dump_register: SP_ELX:    0x40374060
> [CPU1] up_dump_register: TPIDR_EL0: 0x40378700
> [CPU1] up_dump_register: TPIDR_EL1: 0x40378700
> [CPU1] up_dump_register: EXE_DEPTH: 0x1
> [CPU1] dump_tasks:    PID GROUP   CPU PRI POLICY   TYPE    NPX STATE   EVENT      SIGMASK          STACKBASE  STACKSIZE      USED   FILLED    COMMAND
> [CPU1] dump_tasks:   ----   ---  ---- --- -------- ------- --- ------- ---------- -------- 0x402cb000      8192      8192   100.0%!   irq
> [CPU1] dump_task:       0     0     0   0 FIFO     Kthread N-- Assigned           0000000000000000 0x402d1010     16368      1360     8.3%    CPU0 IDLE
> [CPU1] dump_task:       1     1     1   0 FIFO     Kthread N-- Assigned           0000000000000000 0x402d5010     16368      1088     6.6%    CPU1 IDLE
> [CPU1] dump_task:       2     2     2   0 FIFO     Kthread N-- Running            0000000000000000 0x402d9010     16368      1088     6.6%    CPU2 IDLE
> [CPU1] dump_task:       3     3     3   0 FIFO     Kthread N-- Running            0000000000000000 0x402dd010     16368      1088     6.6%    CPU3 IDLE
> [CPU1] dump_task:       4     4     0 192 RR       Kthread --- Waiting Semaphore  0000000000000000 0x402e39f0     16320      1264     7.7%    hpwork 0x402c64f8
> [CPU1] dump_task:       5     5     0 100 RR       Task    --- Waiting Semaphore  0000000000000000 0x402e8000     16336      2944    18.0%    nsh_main
> [CPU1] dump_task:      22    22     0 100 RR       Task    --- Running            0000000000000000 0x4036c6e0     16288      2320    14.2%    ostest Arg1 Arg2 Arg3 Arg4
> [CPU1] dump_task:     155   155     1 100 RR       Task    --- Running            0000000000000000 0x403706c0     16336      3552    21.7%    waiter
> [CPU1] dump_task:      15    15     0 100 RR       Task    --- Waiting Semaphore  0000000000000000 0x40379530     16336      1984    12.1%    ostest
> ```
> 
> Could you please confirm if the latest upstream works with qemu?

ok, I'm fixing the problem, should be fixed today

**Pull request merged**